mesonbuild/modules/unstable_wayland.py

```diff
@@ -0,0 +1,133 @@
+
+import os
+
+from . import ExtensionModule, ModuleReturnValue
+from ..build import CustomTarget
+from ..dependencies import PkgConfigDependency
+from ..interpreterbase import FeatureNew
+from ..mesonlib import File, MesonException
+
+
+class WaylandModule(ExtensionModule):
+
+    @FeatureNew('wayland module', '0.62.0')
+    def __init__(self, interpreter):
+        super().__init__(interpreter)
+
+        self.protocols_dep = None
+        self.pkgdatadir = None
+        self.scanner_bin = None
+
+        self.methods.update({
+            'scan_xml': self.scan_xml,
+            'find_protocol': self.find_protocol,
+        })
+
+    def scan_xml(self, state, args, kwargs):
+        if len(args) == 0 or not all(isinstance(arg, (File, str)) for arg in args):
+            raise MesonException('scan requires at least one string or file argument: the xml file.')
+
+        scope = kwargs.get('scope', 'private')
+        if 'private' != scope != 'public':
+            raise MesonException(f'Invalid scope {scope}.')
+
+        side = kwargs.get('side', 'client')
+        if 'client' != side != 'server':
+            raise MesonException(f'Invalid side {side}.')
+
+        if self.scanner_bin is None:
+            self.scanner_bin = state.find_program('wayland-scanner')
+            if self.scanner_bin is None:
+                raise MesonException("Failed to find wayland-scanner")
+
+        targets = []
+        for arg in args:
+            if isinstance(arg, str):
+                if os.path.isabs(arg):
+                    xml_path = arg
+                else:
+                    xml_path = os.path.join(self.interpreter.environment.source_dir, arg)
+                    if not os.path.exists(xml_path):
+                        raise MesonException(f'The file {xml_path} does not exist.')
+            else:
+                xml_path = arg.absolute_path(
+                    self.interpreter.environment.source_dir,
+                    self.interpreter.environment.build_dir)
+
+            xml_file = os.path.basename(xml_path)
+            name, _ = os.path.splitext(xml_file)
+
+            code = CustomTarget(
+                f'{name}-protocol',
+                state.subdir,
+                state.subproject,
+                [self.scanner_bin, f'{scope}-code', '@INPUT@', '@OUTPUT@'],
+                [xml_path],
+                [f'{name}-protocol.c'],
+                backend=state.backend,
+            )
+            targets.append(code)
+
+            header = CustomTarget(
+                f'{name}-{side}-protocol',
+                state.subdir,
+                state.subproject,
+                [self.scanner_bin, f'{side}-header', '@INPUT@', '@OUTPUT@'],
+                [xml_path],
+                [f'{name}-{side}-protocol.h'],
+                backend=state.backend,
+            )
+            targets.append(header)
+
+        return ModuleReturnValue(targets, [targets])
+
+    def find_protocol(self, state, args, kwargs):
```

Comment:
Should use @typed_pos_args and @typed_kwargs here. They replace the manual type checks you do below.